### ISSUE TITLE
Ultisnips snippet doesn't expand when editing LaTeX files

### ISSUE THREAD

**Issue by TornaxO7** [nvim-treesitter/nvim-treesitter#1184]
**Describe the bug**

When I'm editing a `.tex` file and trying to use one of my auto-snippets, they don't expand.
Here's a live-demo:

https://user-images.githubusercontent.com/50843046/114637424-9d848780-9cc9-11eb-88f9-c2f6376354ff.mp4

**To Reproduce**

1. Put this into your `init.vim`:
<details>
<code>

```vim
call plug#begin('~/.vim/plugged')
Plug 'lervag/vimtex'
Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'} 
call plug#end()
lua require("treesitter")
```
</details>
</code>

2. Put this into your `~/.config/nvim/lua/treesitter.lua`:

<details>
<code>

```lua
local treesitter = require('nvim-treesitter.configs')

treesitter.setup {
    -- one of "all", "maintained" (parsers with maintainers), or a list of languages
    ensure_installed = "maintained",

    highlight = {
        enable = true,  -- false will disable the whole extension
        disable = {},  -- list of language that will be disabled
    },

    indent = {
        enable = false,
    },

    rainbow = {
        enable = false,
    },
}
```

</details>
</code>

3. This snippet into `~/.config/nvim/UltiSnips/tex.snippets`

<details>
<code>

```snippets
global !p
def math():
	return vim.eval('vimtex#syntax#in_mathzone()') == '1'
endglobal

context "math()"
snippet sum "\sum\limits^{}_{}" wA
\sum\limits^{$1}_{$2}$0
endsnippet
```

</details>
</code>

4. Open any `.tex` file. You can use my minimal example for that:

<details>
<code>

```tex
\documentclass[a4paper, 12pt]{scrartcl}

% =============
% Packages 
% =============
\usepackage[ngerman]{babel}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}				% west european chars
\usepackage{amsmath, amsthm, amssymb} 	% math symbols

% =============
% Commands 
% =============

% =================
% Environments 
% =================

% =============
% Settings 
% =============

% =============
% Document 
% =============
\title{Title}
\author{Author}
\date{\today}

\parskip5pt
\parindent0pt

% ------------
% Content 
% ------------
\begin{document}
\begin{align*}
    su%%
\end{align*}
\end{document}
```

</details>
</code>

5. Go to the end of the word `su` (it's between the align environment; I marked
the position with `%%`) and press `m`.

**Expected behavior**

My `sum` snippet should expand automatically but it doesn't unless I disable nvim-treesitter for LaTeX files.

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

```
health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found  0.19.4 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v15.14.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl" }

## Parser/Features H L F I
  - c_sharp        ✓ . . . 
  - svelte         ✓ . ✓ ✓ 
  - beancount      ✓ . ✓ . 
  - c              ✓ ✓ ✓ ✓ 
  - teal           ✓ ✓ ✓ ✓ 
  - java           ✓ ✓ . ✓ 
  - python         ✓ ✓ ✓ ✓ 
  - sparql         ✓ ✓ ✓ ✓ 
  - comment        ✓ . . . 
  - dart           ✓ ✓ . ✓ 
  - lua            ✓ ✓ ✓ ✓ 
  - zig            ✓ ✓ ✓ ✓ 
  - ocaml          ✓ ✓ ✓ . 
  - go             ✓ ✓ ✓ ✓ 
  - bibtex         ✓ . ✓ ✓ 
  - json           ✓ ✓ ✓ ✓ 
  - jsdoc          ✓ . . . 
  - ledger         ✓ . ✓ ✓ 
  - php            ✓ ✓ ✓ ✓ 
  - julia          ✓ ✓ . . 
  - supercollider  ✓ ✓ ✓ ✓ 
  - latex          ✓ . ✓ . 
  - r              ✓ ✓ . . 
  - turtle         ✓ ✓ ✓ ✓ 
  - devicetree     ✓ ✓ ✓ ✓ 
  - toml           ✓ ✓ ✓ ✓ 
  - typescript     ✓ ✓ ✓ ✓ 
  - gdscript       ✓ ✓ . . 
  - nix            ✓ ✓ ✓ . 
  - jsonc          ✓ ✓ ✓ ✓ 
  - regex          ✓ . . . 
  - verilog        ✓ ✓ ✓ . 
  - ql             ✓ ✓ . . 
  - fennel         ✓ ✓ . . 
  - rst            ✓ ✓ . . 
  - query          ✓ ✓ ✓ ✓ 
  - cpp            ✓ ✓ ✓ ✓ 
  - clojure        ✓ ✓ . . 
  - kotlin         ✓ . . . 
  - ruby           ✓ ✓ ✓ ✓ 
  - vue            ✓ . ✓ . 
  - ocamllex       ✓ . . . 
  - html           ✓ ✓ ✓ ✓ 
  - rust           ✓ ✓ ✓ ✓ 
  - bash           ✓ ✓ ✓ . 
  - graphql        ✓ . . ✓ 
  - javascript     ✓ ✓ ✓ ✓ 
  - css            ✓ . ✓ ✓ 
  - glimmer        ✓ . . . 
  - erlang         . . . . 
  - ocaml_interface✓ ✓ ✓ . 
  - tsx            ✓ ✓ ✓ ✓ 

 Legend: H[ighlight], L[ocals], F[olds], I[ndents]
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

</code>
</details>

**Output of `nvim --version`**

<details>
<code>

```
NVIM v0.5.0-dev+1176-g9481310c2
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/cc -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/tornax/.config/yay/build/neovim-nightly-git/src/build/config -I/home/tornax/.config/yay/build/neovim-nightly-git/src/neovim-nightly-git/src -I/usr/include -I/home/tornax/.config/yay/build/neovim-nightly-git/src/build/src/nvim/auto -I/home/tornax/.config/yay/build/neovim-nightly-git/src/build/include
Compiled by tornax

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

</code>
</details>

**Comment by clason** [comment]
Yes, enabling tree-sitter disables the builtin syntax file, which Ultisnips seems to rely on. You can try enabling both at the same time with the option `additional_vim_regex_highlighting = true` (see https://github.com/nvim-treesitter/nvim-treesitter/pull/1042), but this may break other things.

(Note that tree-sitter is still experimental; for production use I would stick with vimtex's much more extensive syntax support for now.)

**Comment by theHamsta** [comment]
It would be something new to me that ultisnips relies on syntax. It works in general on my machine. But since ultisnips can be arbitrary python scripts they can use  any code they want to. Are you sure it's related to `synatx`

**Comment by TornaxO7** [comment]
Currently my workaround is that I add `'latex'` to the `disable` 'array' (I don't know if you call it like that in lua).

**Comment by clason** [comment]
@theHamsta this relies on the vimtex syntax file:
```vim
def math():
	return vim.eval('vimtex#syntax#in_mathzone()') == '1'
endglobal
```
But the issue may not be related to tree-sitter but to the fact that vimtex 2.0 made massive changes to the syntax file (to the better) that are not fully compatible. In particular, that function may be deprecated, see https://github.com/lervag/vimtex/issues/2019

**Comment by theHamsta** [comment]
If the snippets come from context directly we could maybe ask whether they would consider an alternative tree-sitter implementation of those functions when community provides them. We should be able to provide functions like `in_mathzone` either in a plugin or for vmtex directly implemented with tree-sitter

**Comment by theHamsta** [comment]
This would be a implementation for `in_comment`/`in_mathzone` that has no nvim-treesitter dependency. Also avoided queries since we are only want information about the current node and it's parents. Maybe this can be used to fix the snippet. Where can I find the code for the snippet apart from @lervags personal config?
```lua
local has_treesitter, ts = pcall(require, 'vim.treesitter')
local _, query = pcall(require, 'vim.treesitter.query')

local M = {}

local MATH_ENVIRONMENTS = {
  displaymath = true,
  eqnarray = true,
  equation = true,
  math = true,
  array = true,
}
local MATH_NODES = {
  displayed_equation = true,
  inline_formula = true,
}

local function get_node_at_cursor()
  local cursor = vim.api.nvim_win_get_cursor(0)
  local cursor_range = { cursor[1] - 1, cursor[2] }
  local buf = vim.api.nvim_get_current_buf()
  local ok, parser = pcall(ts.get_parser, buf, 'latex')
  if not ok or not parser then return end
  local root_tree = parser:parse()[1]
  local root = root_tree and root_tree:root()

  if not root then return end

  return root:named_descendant_for_range(cursor_range[1], cursor_range[2], cursor_range[1], cursor_range[2])
end

function M.in_comment()
  if has_treesitter then
    local node = get_node_at_cursor()
    while node do
      if node:type() == 'comment' then
        return true
      end
      node = node:parent()
    end
    return false
  end
end

function M.in_mathzone()
  if has_treesitter then
    local buf = vim.api.nvim_get_current_buf()
    local node = get_node_at_cursor()
    while node do
      if MATH_NODES[node:type()] then
        return true
      end
      if node:type() == 'environment' then
        local begin = node:child(0)
        local names = begin and begin:field('name')

        if names and names[1] and MATH_ENVIRONMENTS[query.get_node_text(names[1], buf):gsub('[%s*]', '')] then
          return true
        end
      end
      node = node:parent()
    end
    return false
  end
end

return M
```

> (I don't believe vimtex will ever use tree-sitter since it has to support both vim and neovim.)

I have seen a lot of plugins for both vim/neovim that just use tree-sitter as a opt-in feature if it's available. But for LaTeX, most of the features can be implemented cross-language tree-sitter plugins or with the LaTeX-LSP. VimTeX really does a lot of stuff..

**Comment by TornaxO7** [comment]
> Where can I find the code for the snippet apart from @lervags personal config?

@theHamsta which snippets do you mean?

**Comment by theHamsta** [comment]
The snippets that don't work for you. I was assuming the ones you are using are depending on the snippet that @clason posted (that use vimtex syntax API

**Comment by clason** [comment]
> I have seen a lot of plugins for both vim/neovim that just use tree-sitter as a opt-in feature if it's available. But for LaTeX, most of the features can be implemented cross-language tree-sitter plugins or with the LaTeX-LSP. VimTeX really does a lot of stuff..

Well, I should have qualified "with reasonable effort". Yes, VimTeX does a lot of stuff (and does it well), so there's a large maintenance surface already. Adding special cases will not help in this regard... (If there's a vanilla vim tree-sitter plugin that is half as good as neovim/nvim-treesitter, we can talk again ;))

(And the snippet is in the OP once you expand the `Details` under Step 3, isn't it?)

In any case, thanks for the treesitter snippet!

**Comment by theHamsta** [comment]
There could be a supplement plugin for the latex parser bundling stuff like this.

**Comment by clason** [comment]
@theHamsta Oh, for sure, that's a job for "Nvimtex". I probably misread your reference to VimTeX; my point was just that I wouldn't wait for or expect the maintainer of VimTeX to add these features -- someone else would have to create that plugin from scratch. (I won't do it due to lack of time and skills, and I certainly don't expect you to do it, either!)

**Comment by TornaxO7** [comment]
Ok so I disabled nvim-treesitter now for my LaTeX stuff for some days now and I don't really think that I need treesitter for my LaTeX files. VimTeX is doing a great job so I think that I can close this issue.

**Comment by theHamsta** [comment]
Also integrating into vimtex would be an option. The `autoload/syntax.vim` module only has three functions from which only the `in_comment`, `in_mathzone` are actually used in other modules like completion and movements (it's the in which context-am-I module). tree-sitter only deactivates the syntax from this point of view users might nvim-treesitter not break other parts of vimtex that don't depend on syntax apart from the context recognition.

For me personally, this would be not important because I use LSP for completion and nvim-treesitter-textobjects for textobjects and movements.

The example I was referring to was matchup.vim. They only use nvim-treesitter for Python, the Lua function have the same signature as their regex-based ones so they just use the ones that are available.

**Comment by medwatt** [comment]
I just came across this issue and had no idea what was causing it. I had to disable my plugins on after the other before I found out that treesitter was causing the issue.

**Comment by theHamsta** [comment]
Where does everyone get this snippet from? Maybe the code could just be changed to `if mathzone or syntax==off`

**Comment by clason** [comment]
No, I think nvim-treesitter needs a big, fat disclaimer that it will disable regex syntax _and everything that relies on it_.

(But who am I kidding -- nobody will read it and this will change nothing...)

The real problem is that people will just blindly copy config from something assuming that they need the latest and greatest without informing themselves (at the source!) first.

**Comment by anstadnik** [comment]
Just in case someone will use the snippet for detecting math environments, it's not fully complete, as it doesn't use all math envs. I've changed it the following way:
```lua
local MATH_ENVIRONMENTS = {
    -- math = true,
    displaymath = true,
    equation = true,
    multline = true,
    eqnarray = true,
    align = true,
    array = true,
    split = true,
    alignat = true,
    gather = true,
    flalign = true
}
```
Since I don't really understand what does `math` stand for it's commented.

**Comment by anstadnik** [comment]
Hi stranger!
Seems that you're trying to follow the path of the legendary @gillescastel, just like I do.
Perhaps you used the above snippet for testing for the math env using tree-sitter for the snippets expansion.
After #2436 you might have noticed that the code above stopped working.
For me the following fixes help:
```
function M.in_mathzone()
    if has_treesitter then
        local buf = vim.api.nvim_get_current_buf()
        local node = get_node_at_cursor()
        while node do
            if MATH_NODES[node:type()] then return true end
            if node:type() == 'generic_environment' then
                local begin = node:child(0)
                local names = begin and begin:field('name')

                if names and names[1] and
                    MATH_ENVIRONMENTS[query.get_node_text(names[1], buf):gsub(
                        '{(%w+)%s*%*?}', '%1')] then return true end
            end
            node = node:parent()
        end
        return false
    end
end
```

I'd appreciate it if someone would find a better way of doing that (:

**Comment by mathjiajia** [comment]
Thank you  @anstadnik 

I also notice that there are some changes in the `comment`.
Below should work:

``` lua
local COMMENT = {
	comment = true,
	line_comment = true,
	block_comment = true,
	comment_environment = true,
}

function M.in_comment()
	local node = get_node_at_cursor()
	while node do
		if COMMENT[node:type()] then
			return true
		end
		node = node:parent()
	end
	return false
end
```

**Comment by medwatt** [comment]
@mathjiajia, how do we use these functions  within ultisnips?

**Comment by mathjiajia** [comment]
> @mathjiajia, how do we use these functions within ultisnips?

Currently, I am using `LuaSnip`. So I am afraid I cannot give a good answer here.
Also, notice the update of my comment above.
Maybe this one can help you (not so sure): https://github.com/DukMastaaa/dotfiles/blob/f4196bea5a01992ca2abecc4cdedb54e16dd089f/.config/nvim/UltiSnips/tex.snippets#L13-L17

**Comment by medwatt** [comment]
@mathjiajia, thanks for the link. I'll check it out. BTW, how do you compare luasnip to ultisnips in terms of capabilities? I wanted to give luasnip a try, but was put off by how time consuming and complicated it is to write a write a snippet compared to ultisnips. Is there a good reason to use luasnip instead of ultisnips?

**Comment by mathjiajia** [comment]
@medwatt 
Maybe try with this plugin: https://github.com/smjonas/snippet-converter.nvim

For the advantages, you may check this snippet of mine:
https://github.com/mathjiajia/config.nvim/blob/dc6ac7102e1e2de840544366070c39c889728bcd/luasnippets/tex/phrases.lua#L16
When I jump to the second placeholder, there will be a `telescope bibtex` prompt.
